src/libraries/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/SharedPerformanceCounter.cs

```diff
@@ -285,22 +285,22 @@ private int CalculateMemoryNoBoundsCheck(int oldOffset, int totalSize, out int a
             newCategoryEntryPointer->NextCategoryOffset = 0;
             newCategoryEntryPointer->FirstInstanceOffset = (int)((long)newInstanceEntryPointer - _baseAddress);
             newCategoryEntryPointer->CategoryNameOffset = (int)(nextPtr - _baseAddress);
-            SafeMarshalCopy(_categoryName, (IntPtr)nextPtr);
+            SafeMarshalCopy(_categoryName, checked((IntPtr)nextPtr));
```

Comment:
The checked arithmetic is a bug here. It would fail in 3GB or 4GB address space processes.

Change `long nextPtr` to `byte* nextPtr` instead.